### DISCUSSION TITLE
PWM Laser or PPI Laser

### DISCUSSION THREAD

**Discussion by andy0983** [Paciente8159/uCNC#662] [Q&A]
Hello, I'm glad that the new version includes NUCLEO-F411RE :)
I have a question, I would like to connect a laser to my board, but I don't really know how.
https://neje.shop/products/40w-laser-module-laser-head-for-cnc-laser-cutter-engraver-woodworking-machine
Whether to configure ucnc as PWM Laser or PPI Laser and what do you think about this laser?

**Comment by Paciente8159** [comment]
Hi there,

About the laser this is just a personal opinion, not a recommendation or nothing like that. Will be great for engraving stuff and will be able to cut a lot faster then more cheap 5 or 10W lasers for sure. It's not as powerfull as CO2 laser machine, but should be a very capable diode laser. If it meets your needs in terms of what you plan and need it for that is the most important thing. Is what you intent to do in terms of materials. Soft woods and some hard woods (cut and engrave fine)...metals (light engraving with some limitations).

About laser PWM and Laser PPI well depends on how you adapt better and the application. For cutting is almost all the same. For engraving not so much.

Laser PWM is the normal/classic wait to use it. You need to find a balance of power and speed to get the desired results to a given material.
For example engraving MDF or pine at the same power and speed produce different results.
Engraving MDF at one speed and then at 2x that speed also produces different results.

Laser PPI reduces those 2 variable to a single one. That is exposure time or the time the laser is on per inch.
That means that no matter the speed you run the machine the amount of time the laser will be on by distance will be the same. I just need  to find how much microseconds I want the laser on over your material surface. I can send the machine fast or slow it will produce very similar results.

By this reason I usually use Laser PPI. But you can try both and decide by your self.

**Comment by andy0983** [comment]
I ordered a laser, but it won't be delivered until the end of the week.
The laser needs a PWM/TTL control signal and air switching.
The tests were made using the LaserGRBL program.

I tested what signal was output using the LASER PWM and LASER PPI settings.
LASER PWM - a signal from 0 to 3.3v appeared on this PWM0 pin
LASER PPI - nothing appears on this PWM0 pin, it is 0v all the time

Configure settings

#define TOOL1 spindle_pwm
#define SPINDLE_PWM PWM2
#define SPINDLE_PWM_DIR DOUT3
#define SPINDLE_PWM_COOLANT_FLOOD DOUT0
#define SPINDLE_PWM_COOLANT_MIST DOUT1
#define TOOL2 laser_ppi
#define ENABLE_LASER_PPI
#define LASER_PPI PWM0
#define LASER_PPI_AIR_ASSIST DOUT2
///////////////////////////////////////////////// /////////////////////////////////
#define TOOL_COUNT 2
#define BAUDRATE 115200
#define BOARD BOARD_NUCLEO_F411RE_SHIELD_V3
#define BOARD_NAME "STM32 Nucleo F411RE"
#define PWM0_BIT 9
#define PWM0_PORT A
#define PWM0_CHANNEL 2

**Reply by Paciente8159** [reply]
OK. A couple of pointers:

- You don't need to change PWM pins between tools. You can make both tools share PWM0 pin for example. Each tool as it's own startup code. And both those tools configure the tool pins. That mean that they can modify the pin behavior from simple output pin to PWM and vice versa
- Also note that PPI tool requires some extra MCodes before you start running code [M126, M127 and M128](https://github.com/Paciente8159/uCNC/blob/master/uCNC/src/hal/tools/tools/README.md#laser_ppi).

I recommend playing around with the settings and modes but here are a few tips.

- Choose a mode (M126) that give the effect you prefer (usually I use mode 3 (mixed mode)
- Choose a resolution that matches the line resolution for better results (for example if you are engraving at 5 lines per mm that translates to 127dpi (25.4mm x 5). Use the same value for M127.
- M128 is probably the only value you will need to figure out for each material. Your laser is powerfull. Start with a conservative value (for example 500µs) and burn a pattern (a filled square for example) at the max S value. If it's to light increase the value. If it's overburn lower the value.

For cutting just use PWM mode it's simpler

**Comment by andy0983** [comment]
I've checked these settings before with M126 -128 and nothing.
Laser PPI does not work.
The PPI output is always turned off, i.e. the pin is always 0v and there is no frequency
the same pin with the PWM laser settings works, the voltage changes and the frequency is about 7.89 kHz with the set 8 kHz

I'm sending a piece of gcode

M126 P3
M127 P127
M128 P500
G90 (use absolute coordinates)
G0 X3 Y3 S0
G1 M4 F1000
F1000
Y3 S0
X13 S100
X23 S200
X33 S300
X43 S400
X53 S500
X63 S600
X73 S700
X83 S800
X93 S900
X103 S1000
Y3.125 S0
X93 S1000

**Comment by Paciente8159** [comment]
Send me your whole project and I will check it out.

**Comment by andy0983** [comment]
I am sending the project
[uCNC.zip](https://github.com/Paciente8159/uCNC/files/14706189/uCNC.zip)

**Reply by Paciente8159** [reply]
Just by giving a quick look I spotted some configuration problems.
For once Laser PPI makes use of the oneshot timer. Currently that timer is assigned to TIMER1.  But so is PWM0. 

**ONESHOT_TIMER must be a dedicated timer.**

If you only plan on using PPI mode you can configure the LASER PPI control pin as a generic DOUT pin. It does not have to be a PWM pin.

But if you plan to use the laser module also in PWM mode then keep the configuration like that. You can add an 3rd tool and assign it to laser_pwm using the same pins as the pins used by the laser_ppi tool. By switching between these 2 tools the laser will automatically change working mode from PPI to PWM and vice versa.

**Comment by andy0983** [comment]
You might also want to check for other configuration errors.

I changed it to:

#define ONESHOT_TIMER 10
#define TOOL_COUNT 3
#define TOOL3 laser_pwm

If I want to set the right tool, this is how I do it
M6 T1 - spindle_pwm
M6 T2 - laser_ppi
M6 T3 - laser_pwm

or so

M6 T1 - spindle_pwm

M126 P3       
M127 P127          --- laser_ppi
M128 P1500  

M6 T3 - laser_pwm

**Reply by Paciente8159** [reply]
PPI mode and PPI tool needs a bit of refactoring.

If might not be completely explicit but it's actually possible to change to PPI mode with the M126 command (this command actually modifies settings $32 internally). This minimized need changes to several internal structs and saved some RAM.

What prevents PPI to work properly just via M126 command is the need to configure the laser control pin to a generic output pin (in pwm mode does not work well).

If you look at [PWM tool startup code](https://github.com/Paciente8159/uCNC/blob/970d920bd0c73695f8ffbc07490a1121a39f987b/uCNC/src/hal/tools/tools/laser_pwm.c#L54) on load the tool is set to laser mode (the same as M126 P0)
Also [loading the PPI tool](https://github.com/Paciente8159/uCNC/blob/970d920bd0c73695f8ffbc07490a1121a39f987b/uCNC/src/hal/tools/tools/laser_ppi.c#L247) modifies the tool mode via the tool startup code.

so this does not work at all

```
M126 P3 ;sets PPI motion mode to PPI blend mode
M127 P127 ;sets PPI to 127 PPI
M128 P1500 ;laser on for 1500µs

M6 T3 - laser_pwm 
```

The logic thing is first load the tool for the job and then send commands for the tool 

```
M6 T2 ;load laser_ppi tool
M126 P3 ;sets PPI motion mode to PPI blend mode
M127 P127 ;sets PPI to 127 PPI
M128 P1500 ;laser on for 1500µs

;run your code for the PPI mode

M6 T3 ;load laser_pwm tool. this also disables M126

;run your code for laser PWM mode

```
